src/System.Composition.Convention/src/Microsoft/Internal/Lock.Reader.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+
+namespace Microsoft.Internal
+{
+    internal struct ReadLock : IDisposable
+    {
+        private readonly Lock _lock;
+        private int _isDisposed;
+
+        public ReadLock(Lock @lock)
+        {
+            _isDisposed = 0;
+            _lock = @lock;
+            _lock.EnterReadLock();
+        }
+
+        public void Dispose()
+        {
+            if (Interlocked.CompareExchange(ref _isDisposed, 1, 0) == 0)
```

Comment:
FWIW, I wasn't actually suggesting adding more cost by using a class.  Seems like you don't need the type at all, as:
```
_lock.EnterReadLock();
try { 
    _memberInfos.TryGetValue(memberInfo, out cachedAttributes);
}
finally { _lock.ExitReadLock(); }
```
seems just as clear as:
```
using(new ReadLock(_lock))
{
    _memberInfos.TryGetValue(memberInfo, out cachedAttributes);
}
```
and without any additional overhead.  Especially since there are only a handful of places internally where these are used.  If you really wanted to keep the struct, because you believe it does add to the simplicity and there are no concerns about misuse, then the Interlocked itself seems like unnecessary overhead and could just be removed: if we're relying on the consumer only using the right pattern, then trying to protect against one particular misuse doesn't seem warranted.